File: frontend/src/api/notifications.ts
import { API_BASE_URL, getApiErrorMessage } from '../config/api';

export type ChannelType = 'email' | 'sms';
export type TriggerType = 'status_change';

export interface NotificationChannel {
  id: number;
  name: string;
  channel_type: ChannelType;
  config: Record<string, unknown>;
  enabled: boolean;
  created_at: string;
}

export interface NotificationChannelCreate {
  name: string;
  channel_type: ChannelType;
  config: Record<string, unknown>;
  enabled?: boolean;
}

export interface NotificationChannelUpdate {
  name?: string;
  channel_type?: ChannelType;
  config?: Record<string, unknown>;
  enabled?: boolean;
}

export interface NotificationRule {
  id: number;
  name: string;
  machine_id: number | null;
  trigger_type: TriggerType;
  trigger_config: Record<string, unknown>;
  channel_ids: number[];
  enabled: boolean;
  created_at: string;
}

export interface NotificationRuleCreate {
  name: string;
  machine_id: number | null;
  trigger_type: TriggerType;
  trigger_config: Record<string, unknown>;
  channel_ids: number[];
  enabled?: boolean;
}

export interface NotificationRuleUpdate {
  name?: string;
  machine_id?: number | null;
  trigger_type?: TriggerType;
  trigger_config?: Record<string, unknown>;
  channel_ids?: number[];
  enabled?: boolean;
}

export interface NotificationLogEntry {
  id: number;
  rule_id: number | null;
  rule_name?: string | null;
  channel_id: number | null;
  machine_id: number | null;
  event_type: string;
  event_data: Record<string, unknown> | null;
  message: string | null;
  status: string;
  error_message: string | null;
  sent_at: string;
}

export interface TestResult {
  status: 'sent' | 'failed';
  error?: string;
}

// --- Channels ---

export async function listChannels(): Promise<NotificationChannel[]> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/channels`);
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

export async function createChannel(data: NotificationChannelCreate): Promise<NotificationChannel> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/channels`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

export async function updateChannel(id: number, data: NotificationChannelUpdate): Promise<NotificationChannel> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/channels/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

export async function deleteChannel(id: number): Promise<void> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/channels/${id}`, { method: 'DELETE' });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
}

export async function testChannel(id: number): Promise<TestResult> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/test/${id}`, { method: 'POST', headers: { 'Content-Type': 'application/json' } });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

// --- Rules ---

export async function listRules(): Promise<NotificationRule[]> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/rules`);
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

export async function createRule(data: NotificationRuleCreate): Promise<NotificationRule> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/rules`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

export async function updateRule(id: number, data: NotificationRuleUpdate): Promise<NotificationRule> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/rules/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

export async function deleteRule(id: number): Promise<void> {
  const res = await fetch(`${API_BASE_URL}/api/notifications/rules/${id}`, { method: 'DELETE' });
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
}

// --- Log ---

export async function listLog(params?: { machine_id?: number; limit?: number; offset?: number }): Promise<NotificationLogEntry[]> {
  const query = new URLSearchParams();
  if (params?.machine_id != null) query.set('machine_id', String(params.machine_id));
  if (params?.limit != null) query.set('limit', String(params.limit));
  if (params?.offset != null) query.set('offset', String(params.offset));
  const res = await fetch(`${API_BASE_URL}/api/notifications/log?${query}`);
  if (!res.ok) throw new Error(await getApiErrorMessage(res));
  return res.json();
}

File: backend/app/api/notifications.py
"""API endpoints for notification channel and rule management."""
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db.base import get_db
from app.models.notification import NotificationChannel, NotificationLog, NotificationRule
from app.schemas.notification import (
    NotificationChannelCreate,
    NotificationChannelResponse,
    NotificationChannelUpdate,
    NotificationLogResponse,
    NotificationRuleCreate,
    NotificationRuleResponse,
    NotificationRuleUpdate,
)
from app.utils.secret_redaction import merge_channel_config_update

router = APIRouter()

_notification_service = None


def set_notification_service(service):
    global _notification_service
    _notification_service = service


# --- Channels ---

@router.get("/channels", response_model=list[NotificationChannelResponse])
def list_channels(db: Session = Depends(get_db)):
    return db.query(NotificationChannel).order_by(NotificationChannel.id).all()


@router.post("/channels", response_model=NotificationChannelResponse, status_code=201)
def create_channel(payload: NotificationChannelCreate, db: Session = Depends(get_db)):
    channel = NotificationChannel(**payload.model_dump())
    db.add(channel)
    db.commit()
    db.refresh(channel)
    return channel


@router.get("/channels/{channel_id}", response_model=NotificationChannelResponse)
def get_channel(channel_id: int, db: Session = Depends(get_db)):
    channel = db.query(NotificationChannel).filter(NotificationChannel.id == channel_id).first()
    if not channel:
        raise HTTPException(status_code=404, detail="Channel not found")
    return channel


@router.put("/channels/{channel_id}", response_model=NotificationChannelResponse)
def update_channel(channel_id: int, payload: NotificationChannelUpdate, db: Session = Depends(get_db)):
    channel = db.query(NotificationChannel).filter(NotificationChannel.id == channel_id).first()
    if not channel:
        raise HTTPException(status_code=404, detail="Channel not found")
    for field, value in payload.model_dump(exclude_unset=True).items():
        if field == "config" and value is not None:
            value = merge_channel_config_update(channel.config, value)
        setattr(channel, field, value)
    db.commit()
    db.refresh(channel)
    return channel


@router.delete("/channels/{channel_id}", status_code=204)
def delete_channel(channel_id: int, db: Session = Depends(get_db)):
    channel = db.query(NotificationChannel).filter(NotificationChannel.id == channel_id).first()
    if not channel:
        raise HTTPException(status_code=404, detail="Channel not found")
    db.delete(channel)
    db.commit()


@router.post("/test/{channel_id}")
async def test_channel(channel_id: int, db: Session = Depends(get_db)):
    channel = db.query(NotificationChannel).filter(NotificationChannel.id == channel_id).first()
    if not channel:
        raise HTTPException(status_code=404, detail="Channel not found")
    if not _notification_service:
        raise HTTPException(status_code=503, detail="Notification service unavailable")

    status, error = await _notification_service.send_test(channel)
    if status == "sent":
        return {"status": "sent"}
    return {"status": "failed", "error": error}


# --- Rules ---

@router.get("/rules", response_model=list[NotificationRuleResponse])
def list_rules(db: Session = Depends(get_db)):
    return db.query(NotificationRule).order_by(NotificationRule.id).all()


@router.post("/rules", response_model=NotificationRuleResponse, status_code=201)
def create_rule(payload: NotificationRuleCreate, db: Session = Depends(get_db)):
    rule = NotificationRule(**payload.model_dump())
    db.add(rule)
    db.commit()
    db.refresh(rule)
    return rule


@router.get("/rules/{rule_id}", response_model=NotificationRuleResponse)
def get_rule(rule_id: int, db: Session = Depends(get_db)):
    rule = db.query(NotificationRule).filter(NotificationRule.id == rule_id).first()
    if not rule:
        raise HTTPException(status_code=404, detail="Rule not found")
    return rule


@router.put("/rules/{rule_id}", response_model=NotificationRuleResponse)
def update_rule(rule_id: int, payload: NotificationRuleUpdate, db: Session = Depends(get_db)):
    rule = db.query(NotificationRule).filter(NotificationRule.id == rule_id).first()
    if not rule:
        raise HTTPException(status_code=404, detail="Rule not found")
    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(rule, field, value)
    db.commit()
    db.refresh(rule)
    return rule


@router.delete("/rules/{rule_id}", status_code=204)
def delete_rule(rule_id: int, db: Session = Depends(get_db)):
    rule = db.query(NotificationRule).filter(NotificationRule.id == rule_id).first()
    if not rule:
        raise HTTPException(status_code=404, detail="Rule not found")
    db.delete(rule)
    db.commit()


# --- Delivery Log ---

@router.get("/log", response_model=list[NotificationLogResponse])
def list_log(
    machine_id: Optional[int] = None,
    limit: int = 50,
    offset: int = 0,
    db: Session = Depends(get_db),
):
    limit = min(limit, 200)
    query = db.query(NotificationLog)
    if machine_id is not None:
        query = query.filter(NotificationLog.machine_id == machine_id)
    rows = (
        query.order_by(NotificationLog.sent_at.desc())
        .limit(limit)
        .offset(offset)
        .all()
    )

    rule_ids = sorted({r.rule_id for r in rows if r.rule_id is not None})
    rule_name_by_id = {}
    if rule_ids:
        rule_name_by_id = {
            rule.id: rule.name
            for rule in db.query(NotificationRule.id, NotificationRule.name)
            .filter(NotificationRule.id.in_(rule_ids))
            .all()
        }

    for row in rows:
        setattr(row, "rule_name", rule_name_by_id.get(row.rule_id))

    return rows
